### ISSUE TITLE
Flutter 3.7.12 with native plugin fails with AGP version 8.0+

### ISSUE THREAD

**Issue by maxlapides** [flutter/flutter#125181]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create a new Flutter Application (empty)
2. Use Android Studio's "AGP Upgrade Assistant" to upgrade to Gradle 8
3. `flutter pub add device_info_plus` (or any other package that has a native component like `mobile_scanner`)
4. Gradle sync

### Expected results

Gradle sync success

### Actual results

Gradle sync failure:

```
Namespace not specified. Please specify a namespace in the module's build.gradle file like so:

android {
    namespace 'com.example.namespace'
}

If the package attribute is specified in the source AndroidManifest.xml, it can be migrated automatically to the namespace value in the build.gradle file using the AGP Upgrade Assistant; please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.
```

### Code sample

<details><summary>Code sample</summary>

```dart
name: flutter_application_1
description: A new Flutter project.
publish_to: 'none'
version: 0.1.0

environment:
  sdk: '>=2.19.6 <3.0.0'

dependencies:
  flutter:
    sdk: flutter
  flutter_timezone: ^1.0.4

dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_lints: ^2.0.0

flutter:
  uses-material-design: true
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.7.12, on macOS 13.3.1 22E261 darwin-arm64, locale en-US)
    • Flutter version 3.7.12 on channel stable at /Users/maxlapides/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4d9e56e694 (2 days ago), 2023-04-17 21:47:46 -0400
    • Engine revision 1a65d409c7
    • Dart version 2.19.6
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/maxlapides/Library/Android/sdk
    • Platform android-33, build-tools 33.0.2
    • ANDROID_HOME = /Users/maxlapides/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        • android-arm64  • Android 13 (API 33) (emulator)
    • iPhone 14 Pro Max (mobile)  • 988C1D17-7F2A-48E1-A3D3-3F77767A91D1 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)             • macos                                • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)                • chrome                               • web-javascript • Google Chrome 112.0.5615.121

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by vbuberen** [comment]
In the error there is a clear mention that now you need to add `namespace` in `android` section.

More info in the official docs:
https://developer.android.com/build/releases/gradle-plugin#namespace-dsl
https://developer.android.com/build/configure-app-module#set-namespace

**Comment by MacLeeIn** [comment]
Version 3.7.12 working perfectly for me. AGP version 7.4.2 and Gradle 8.0 

Put your namespace on android/app/build.gradle file android tag

**Comment by huycozy** [comment]
Thanks for the report! I also can see this issue when adding `device_info_plus` plugin. Using `Android Studio Flamingo | 2022.2.1`.  But after adding namespace as the suggestion above, Gradle can sync successfully. 

For anyone facing this issue, please try it to see if it works.

**Comment by Shakle** [comment]
> Version 3.7.12 working perfectly for me. AGP version 7.4.2 and Gradle 8.0
> 
> Put your namespace on android/app/build.gradle file android tag
> 
> https://developer.android.com/build/configure-app-module

You don't have dependencies which use native code

**Comment by ahmedhnewa** [comment]
> Version 3.7.12 working perfectly for me. AGP version 7.4.2 and Gradle 8.0 
> 
> Put your namespace on android/app/build.gradle file android tag 
> 
> https://developer.android.com/build/configure-app-module

Hi, this namespace need to be added on all build.gradle of the plugins

**Comment by Shakle** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the [existing issues](https://github.com/flutter/flutter/issues)
> * [x]  I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)
> 
> ### Steps to reproduce
> 1. Create a new Flutter Application (empty)
> 2. Use Android Studio's "AGP Upgrade Assistant" to upgrade to Gradle 8
> 3. `flutter pub add device_info_plus` (or any other package that has a native component like `mobile_scanner`)
> 4. Gradle sync
> 
> ### Expected results
> Gradle sync success
> 
> ### Actual results
> Gradle sync failure:
> 
> ```
> Namespace not specified. Please specify a namespace in the module's build.gradle file like so:
> 
> android {
>     namespace 'com.example.namespace'
> }
> 
> If the package attribute is specified in the source AndroidManifest.xml, it can be migrated automatically to the namespace value in the build.gradle file using the AGP Upgrade Assistant; please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.
> ```
> 
> ### Code sample
> Code sample
> ### Screenshots or Video
> Screenshots / Video demonstration
> ### Logs
> Logs
> ### Flutter Doctor output
> Doctor output

https://github.com/juliansteenbakker/mobile_scanner/issues/602

mobile scanner will soon be updated

**Comment by stuartmorgan** [comment]
I tested one of our plugins with the `namespace` added but the `package` left in `AndroidManifest.xml`, and built it with an example app that was set to AGP 4.2/Gradle 6.7.1 (which predates `namespace` support according to [the docs](https://developer.android.com/build/releases/gradle-plugin#namespace-dsl), as well as AGP 8.0/Gradle 8.0.0, and both worked.

So it appears that old versions of AGP are fine with `namespace` existing, and new versions are fine with the duplicate entry in `AndroidManfest.xml`, in which case we can easily support AGP 8+ without breaking support for AGP <7, which means this should be straightforward for plugin authors to fix.

On the Flutter side, we'll need to:
- [ ] Update all of the plugins in `flutter/packages` to include the namespace.
- [ ] Update the `integration_test` plugin in `flutter/flutter` to include the namespace.
  - [ ] Ideally we would cherry-pick this to beta and/or stable; since it's used as an SDK-sourced package, apps that depend on `integration_test` will be blocked from updating to 8.0 until they are using a Flutter version that has the addition.
- [ ] Update the plugin template in `flutter/flutter` to include the namespace.

Optional changes:
- [ ] Add tooling support to catch this error and add a Flutter error message telling people to report the issue to the author of the relevant plugins.
- [ ] Add the namespace to the app template in `flutter/flutter` (we won't *need* to do that until we update the AGP version in the template).

**Comment by reidbaker** [comment]
Changed the title since the issue is more closely related to AGP than it is to gradle.

**Comment by stuartmorgan** [comment]
Adjusting title; the breaking change is in AGP 8.0

**Comment by reidbaker** [comment]
Update it is unlikely we will back port AGP 8 support to flutter 3.7. 

Currently working on a fix to backport AGP 8 to 3.10 beta.

**Comment by reidbaker** [comment]
Fixing this bug I think is blocked on https://github.com/flutter/flutter/issues/125325

**Comment by reidbaker** [comment]
After a conversation with @stuartmorgan we might still be able to fix this without a test.

**Comment by reidbaker** [comment]
Namspace appers to be introduced in 7.3.* https://developer.android.com/build/releases/past-releases/agp-7-3-0-release-notes

**Comment by stuartmorgan** [comment]
Apparently my testing was incorrect in some way; doing this step:

> - [ ] Update all of the plugins in `flutter/packages` to include the namespace.

broke a whole lot of people (see https://github.com/flutter/flutter/issues/125607 and https://github.com/flutter/flutter/issues/125605 for example). I'll need to revert that across all plugins, and we'll need to continue investigating whether there's a way to make it work in a backwards-compatible way. If there isn't one, we may need to drop AGP <7.3 support before we can support 8.0.

**Comment by stuartmorgan** [comment]
> Namspace appers to be introduced in 7.3.* https://developer.android.com/build/releases/past-releases/agp-7-3-0-release-notes

Interestingly, in [this comment](https://github.com/flutter/flutter/issues/125607#issuecomment-1525324210) someone reported that they locally fixed breakage from the updates that added `namespace` by updating their project from AGP 4.x to AGP 7.2. So maybe the warning was introduced in 7.3, but support for the attribute came earlier in the 7.x line?

**Comment by ahmedhnewa** [comment]
> Apparently my testing was incorrect in some way; doing this step:
> 
> > - [ ] Update all of the plugins in `flutter/packages` to include the namespace.
> 
> broke a whole lot of people (see https://github.com/flutter/flutter/issues/125607 and https://github.com/flutter/flutter/issues/125605 for example). I'll need to revert that across all plugins, and we'll need to continue investigating whether there's a way to make it work in a backwards-compatible way. If there isn't one, we may need to drop AGP <7 support before we can support 8.0.

Then I suggest that we as developers, drop out gradle 7.3 and older and request from all the developers to migrate to gradle 8, I mean, google will do it sooner or later so this is for the best, and also make flutter project creation command generate the new projects in gradle 8 and use namespace in build.gradle instead of define package attribute in AndroidManifest.xml

**Comment by stuartmorgan** [comment]
> Even though adding of the property broke builds for some people, it might be a good opportunity to update a big bunch of projects which still rely on old versions of AGP (older than 7) to bring them up to date with current state of Android build configs.

It's true that there needs to be a path to get developers onto newer versions of AGP, but "ship updates in all of our plugins that give people incredibly opaque Gradle error messages with no indication of what is wrong or how to fix it" is not a developer-friendly path for doing that. That's why I pre-tested to (try to; apparently unsuccessfully) avoid that outcome.

The right way to get people onto new versions of AGP and Gradle is via `flutter` tool support for that, in the form of warnings or errors from `flutter` itself with very clear, actionable messages. That's certainly something we can explore, but ideally we want that to be decoupled from supporting AGP 8, as the timeline for a toolchain requirement change is longer (e.g., very targeted fixes for AGP 8.0 support could potentially be cherry-picked to 3.10, but dropping 4.x would not be.)

> My colleague on Plus Plugins created a possible workaround for this issue: https://github.com/fluttercommunity/plus_plugins/issues/1722

Thanks, I'll take a look at whether that's (or something similar) is something we could consider adopting in our plugins as well, rather than reverting the namespace addition.

> but I am not optimistic about it as it would allow people to continue ignoring updates that Google brings to Android builds speed and other optimisations introduced regularly with new releases of AGP.

While the choice is obviously up to each plugin's developer how to handle this, I would encourage individual plugin developers to *not* consider it within the scope of their responsibility to force developers to update their toolchains by intentionally leaving them in a broken, hard-to-understand state.

**Comment by stuartmorgan** [comment]
> Apparently my testing was incorrect in some way; doing this step:
> 
> > * [ ]  Update all of the plugins in `flutter/packages` to include the namespace.
> 
> broke a whole lot of people (see #125607 and #125605 for example).

After retesting and doing more investigation, it turns out the problem is that I tested AGP 4.2, but not AGP 4.1, and it doesn't work in 4.1. That's because namespace was actually [present in 4.2](https://developer.android.com/reference/tools/gradle-api/4.2/com/android/build/api/dsl/CommonExtension#namespace) (thanks to @reidbaker for finding that), even though the doc say it was introduced in 7.0 as noted a comment above. It's [not in 4.1](https://developer.android.com/reference/tools/gradle-api/4.1/com/android/build/api/dsl/CommonExtension).

Unfortunately prior to Flutter 3.0, `flutter create` used AGP 4.1 in the build.gradle file it created (it changed to 7.1 in Flutter 3.0; we never created anything with 4.2), so it's very likely that a lot of people on 4.x are on 4.1. (As an added issue, current versions of Android Studio don't even handle AGP 4.1 in the AGP update assistant, so even if a developer opens their project in Android Studio it won't offer to update them; it needs to be done manually for projects that old.)

**Comment by stuartmorgan** [comment]
For any plugin authors following this to determine what to do with their own plugins, https://github.com/flutter/flutter/issues/125621#issuecomment-1525995461 has a conditional version of adding `namespace` that we've verified works in AGP 8.0 without breaking 4.1.

**Comment by mchudy** [comment]
Why official Flutter plugins don't keep strict dependency versions on related platform packages? This issue with introducing a breaking change mostly in Android packages comes back a lot. The fact that specifying a strict version of a plugin in project pubspec.yaml (`path_provider: 2.0.11`) doesn't give us a guarantee that every consecutive CI build will have the same code is really alarming (it can download different version of path_provider_android 2.0.x version on every build). This can not only cause sudden build failures on the CI, but also potentially introduce regressions in a supposedly identical build. In my project, upgrading Gradle is not straightforward due to some corporate proxy restrictions and tracking and overriding dependency versions of every plugin is hardly maintainable.

**Comment by stuartmorgan** [comment]
@mchudy Please file a new issue if you want to propose a different way of managing dependencies; it's unrelated to this issue.

**Comment by reidbaker** [comment]
Final State: 

* AGP 8.0 requires the entire plugin ecosystem to add namespace to their build.gradle files. 
* [This comment](https://github.com/flutter/flutter/issues/125621#issuecomment-1525995461) provides a way for plugin authors to provide a backwards compatible namespace declaration. 
* The root of the issue for AGP 8.0 adoption is based on plugins not based on flutter. The exception to that is if your app uses integration_test which is published as part of flutter. To use AGP 8.0 and run integration_tests you must update to flutter 3.10. 

Closing issue.

**Comment by EphraimNetWorks** [comment]
You can also add this under the `subprojects` block in your root build.gradle (android/build.gradle). It checks if namespace in plugin hasn't been set and sets it using package value from it's AndroidManifest.xml

```groovy
subprojects {
    //other config
    afterEvaluate {
        // check if `android` block is available
        if(it.hasProperty('android') && it.android.namespace == null){
            def manifest = new XmlSlurper().parse(file(it.android.sourceSets.main.manifest.srcFile))
            def packageName = manifest.@package.text()
            print("Setting ${packageName} as android namespace")
            android.namespace= packageName
        }
    }
}

```

**Comment by ahmedhnewa** [comment]
It's simple, you don't even need to edit flutter android plugin (android native code) 
edit your build.gradle (project module) to the following:

```
subprojects {
    project.buildDir = "${rootProject.buildDir}/${project.name}"
    afterEvaluate {
        // check if android block is available

        if (it.hasProperty('android')) {

            if (it.android.namespace == null) {
                def manifest = new XmlSlurper().parse(file(it.android.sourceSets.main.manifest.srcFile))
                def packageName = manifest.@package.text()
                println("Setting ${packageName} as android namespace")
                android.namespace = packageName
            }

            def javaVersion = JavaVersion.VERSION_17
            android {
                def androidApiVersion = 33
                compileSdkVersion androidApiVersion
                defaultConfig {
                    targetSdkVersion androidApiVersion
                }
                compileOptions {
                    sourceCompatibility javaVersion
                    targetCompatibility javaVersion
                }
                tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).configureEach {
                    kotlinOptions {
                        jvmTarget = javaVersion.toString()
                    }
                }
                println("Setting java version to ${javaVersion.toString()} which is $javaVersion")
                println("Setting compileSdkVersion and targetSdkVersion to $androidApiVersion")
            }
        }

    }

}
subprojects {
    project.evaluationDependsOn(':app')
}
```
